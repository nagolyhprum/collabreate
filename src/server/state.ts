export const defaultAdminState = () => ({
    ui : {},
    selectedDirectory : "components",
    files : [],
    components : [],
    branch : {
        id : "",
        previousBranchId : "",
        projectId : ""
    },
    project : {
        id : "",
        latestBranchId : ""
    },
    modal : {
        remove : {
            id : -1,
            name : "",
            input : ""
        },
        rename : {
            id : -1,
            name : "",
            input : ""
        },
        move : {
            id : -1,
            name : "",
            parentId : null
        }
    },
})