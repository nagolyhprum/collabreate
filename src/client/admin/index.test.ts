import { File } from '@prisma/client'
import { defaultAdminState } from '../../server/state'
import { test } from '../components'
import { 
    RenameModal, 
    FileComponent, 
    FolderComponent, 
    RemoveModal, 
    MoveModal
} from './'

describe("Components", () => {
    describe("RenameModal", () => {
        it("requires a name", () => {
            const global = defaultAdminState()
            const document = test(RenameModal, global, global)
            // it gets enabled
            document.input("rename_modal_name_input", "test")
            expect(document.id("rename_modal_save_button")?.enabled).toBe(true)
            // it gets disabled
            document.input("rename_modal_name_input", "")
            expect(document.id("rename_modal_save_button")?.enabled).toBe(false)
        })
        it("requires a unique name", () => {
            const global = defaultAdminState({
                modal : {
                    rename : {
                        id : 1,                        
                    }
                },
                files : [{
                    branchId : "",
                    id : -1,
                    isFolder : false,
                    name : "a",
                    parentId : -1,
                    uiId : ""
                }, {
                    branchId : "",
                    id : -1,
                    isFolder : true,
                    name : "b",
                    parentId : -1,
                    uiId : ""
                }]
            })
            const document = test(RenameModal, global, global)
            document.input("rename_modal_name_input", "a")
            expect(document.id("rename_modal_save_button")?.enabled).toBe(false)
            document.input("rename_modal_name_input", "b")
            expect(document.id("rename_modal_save_button")?.enabled).toBe(false)
            document.input("rename_modal_name_input", "c")
            expect(document.id("rename_modal_save_button")?.enabled).toBe(true)
        })
        it("can open and close", () => {
            const global = defaultAdminState({
                modal : {
                    rename : {
                        id : 1,                        
                    }
                }
            })
            const document = test(RenameModal, global, global)
            expect(document.id("rename_modal")?.visible).toBe(true)
            // lets cancel the rename
            document.click("rename_modal_cancel_button")
            expect(document.id("rename_modal")?.visible).toBe(false)
        })
        it("can rename a folder / file", () => {
            const global = defaultAdminState({
                modal : {
                    rename : {
                        id : 1,
                        input : "test 123",
                        name : "test"
                    }
                }
            })
            const fetch = jest.fn();
            const document = test(RenameModal, global, global, {
                fetch
            })
            // make sure it observes properly
            expect(document.id("rename_modal")?.visible).toBe(true)
            expect(document.id("rename_modal_title")?.text).toBe("Renaming: test")
            // save it
            document.click("rename_modal_save_button")
            expect(fetch).toBeCalledWith("/api/file", {
                method : "PATCH",
                headers : {
                    "Content-Type" : "application/json; charset=utf-8"
                },
                body : JSON.stringify({
                    id : 1,
                    name : "test 123"
                })
            })
            // it should close now
            expect(document.id("rename_modal")?.visible).toBe(false)
        })
    })
    describe("FileComponent", () => {
        const file = (file : Partial<File> = {}) => ({
            branchId : "",
            id : -1,
            isFolder : false,
            name : "",
            parentId : null,
            uiId : "",
            ...file
        })
        it("shows the files name", () => {
            const global = defaultAdminState()
            const local = file({
                name : "test"
            })
            const document = test(FileComponent, global, local)
            expect(document.id("file_component_name")?.text).toBe("test")
        })
        it("can open rename", () => {
            const global = defaultAdminState()
            const local = file({
                id : 1,
                name : "test"
            })
            const document = test(FileComponent, global, local)
            document.click("file_component_rename_buttom")
            expect(global.modal.rename).toEqual({
                id : 1,
                name : "test",
                input : "test"
            })
        })
        it("can open move", () => {
            const global = defaultAdminState()
            const local = file({
                id : 1,
                name : "test",
                parentId : 2
            })
            const document = test(FileComponent, global, local)
            document.click("file_component_move_buttom")
            expect(global.modal.move).toEqual({
                id : 1,
                name : "test",
                parentId : 2
            })
        })
        it("can open remove", () => {
            const global = defaultAdminState()
            const local = file({
                id : 1,
                name : "test"
            })
            const document = test(FileComponent, global, local)
            document.click("file_component_remove_buttom")
            expect(global.modal.remove).toEqual({
                id : 1,
                name : "test",
                input : ""
            })
        })
    })

    describe("FolderComponent", () => {
        const folder = (folder : Partial<File> = {}) => ({
            branchId : "",
            id : -1,
            isFolder : true,
            name : "",
            parentId : null,
            uiId : "",
            ...folder
        })
        it("shows the folders name", () => {
            const global = defaultAdminState()
            const local = folder({
                name : "test"
            })
            const document = test(FolderComponent, global, local)
            expect(document.id("folder_component_name")?.text).toBe("test")
        })
        it("can open rename", () => {
            const global = defaultAdminState()
            const local = folder({
                id : 1,
                name : "test"
            })
            const document = test(FolderComponent, global, local)
            document.click("folder_component_rename_button")
            expect(global.modal.rename).toEqual({
                id : 1,
                name : "test",
                input : "test"
            })
        })
        it("can open move", () => {
            const global = defaultAdminState()
            const local = folder({
                id : 1,
                name : "test",
                parentId : 2
            })
            const document = test(FolderComponent, global, local)
            document.click("folder_component_move_button")
            expect(global.modal.move).toEqual({
                id : 1,
                name : "test",
                parentId : 2
            })
        })
        it("can open remove", () => {
            const global = defaultAdminState()
            const local = folder({
                id : 1,
                name : "test"
            })
            const document = test(FolderComponent, global, local)
            document.click("folder_component_remove_button")
            expect(global.modal.remove).toEqual({
                id : 1,
                name : "test",
                input : ""
            })
        })
        it("can show/hide children", () => {
            const global = defaultAdminState({
                ui : {
                    "" : false
                }
            })
            const local = folder()
            const document = test(FolderComponent, global, local)
            expect(document.id("folder_component_children")?.visible).toBe(false)
            expect(document.id("folder_component_expand_collapse_text")?.text).toBe("+")
            document.click("folder_component_expand_collapse_button")
            expect(document.id("folder_component_children")?.visible).toBe(true)
            expect(document.id("folder_component_expand_collapse_text")?.text).toBe("-")
        })
        it("can add a new file", () => {
            const fetch = jest.fn();
            const global = defaultAdminState({
                ui : {
                    "" : false
                }
            })
            const local = folder()
            const document = test(FolderComponent, global, local, {
                fetch
            })
            expect(document.id("folder_component_children")?.visible).toBe(false)
            document.click("folder_component_add_file_button")
            expect(fetch).toBeCalledWith("/api/file", {
                method : "POST",
                headers : {
                    "Content-Type" : "application/json; charset=utf-8"
                },
                body : JSON.stringify({
                    branchId : "",
                    isFolder : false,
                    parentId : -1,
                })
            })
            expect(document.id("folder_component_children")?.visible).toBe(true)
        })
        it("can add a new folder", () => {
            const fetch = jest.fn();
            const global = defaultAdminState({
                ui : {
                    "" : false
                }
            })
            const local = folder()
            const document = test(FolderComponent, global, local, {
                fetch
            })
            expect(document.id("folder_component_children")?.visible).toBe(false)
            document.click("folder_component_add_folder_button")
            expect(fetch).toBeCalledWith("/api/file", {
                method : "POST",
                headers : {
                    "Content-Type" : "application/json; charset=utf-8"
                },
                body : JSON.stringify({
                    branchId : "",
                    isFolder : true,
                    parentId : -1,
                })
            })
            expect(document.id("folder_component_children")?.visible).toBe(true)
        })
    })
    describe("RemoveModal", () => {
        it("save enabled is properly tracked", () => {
            const global = defaultAdminState({
                modal : {
                    remove : {
                        id : 1,
                        input : "",
                        name : "test"
                    }
                }
            })
            const document = test(RemoveModal, global, global)
            expect(document.id("remove_modal_name_input")?.placeholder).toBe("test")
            // it gets disabled
            document.input("remove_modal_name_input", "")
            expect(document.id("remove_modal_save_button")?.enabled).toBe(false)
            // it stays disabled
            document.input("remove_modal_name_input", "te")
            expect(document.id("remove_modal_save_button")?.enabled).toBe(false)
            // it gets enabled
            document.input("remove_modal_name_input", "test")
            expect(document.id("remove_modal_save_button")?.enabled).toBe(true)
        })
        it("can open and close", () => {
            const global = defaultAdminState({
                modal : {
                    remove : {
                        id : 1,
                        input : "",
                        name : "test"
                    }
                }
            })
            const document = test(RemoveModal, global, global)
            expect(document.id("remove_modal")?.visible).toBe(true)
            document.click("remove_modal_cancel_button")
            expect(document.id("remove_modal")?.visible).toBe(false)
        })
        it("can remove a folder / file", () => {
            const global = defaultAdminState({
                modal : {
                    remove : {
                        id : 1,
                        input : "test",
                        name : "test"
                    }
                }
            })
            const fetch = jest.fn();
            const document = test(RemoveModal, global, global, {
                fetch
            })
            // make sure it observes properly
            expect(document.id("remove_modal")?.visible).toBe(true)
            // save it
            document.click("remove_modal_save_button")
            expect(fetch).toBeCalledWith("/api/file", {
                method : "DELETE",
                headers : {
                    "Content-Type" : "application/json; charset=utf-8"
                },
                body : JSON.stringify({
                    id : 1
                })
            })
            // it should close now
            expect(document.id("remove_modal")?.visible).toBe(false)
        })
    })
    describe("MoveModal", () => {
        it("opens and closes", () => {
            const global = defaultAdminState({
                modal : {
                    move : {
                        id : 1,
                        parentId : null,
                        name : "test"
                    }
                }
            })
            const document = test(MoveModal, global, global)
            expect(document.id("move_modal_title")?.text).toBe("Where would you like to move test")
            expect(document.id("move_modal")?.visible).toBe(true)
            document.click("move_modal_cancel_button")
            expect(document.id("move_modal")?.visible).toBe(false)
        })
        it("can update the parent", () => {
            const global = defaultAdminState()
            const document = test(MoveModal, global, global)
            document.select("move_modal_parent_select", "null")
            expect(global.modal.move.parentId).toBe(null)
            document.select("move_modal_parent_select", "1")
            expect(global.modal.move.parentId).toBe(1)
        })
        it("can save, which closes modal and expands folder", () => {
            const fetch = jest.fn()
            const global = defaultAdminState({
                files : [{
                    id : 2,
                    branchId : "",
                    isFolder : true,
                    name : "name",
                    parentId : null,
                    uiId : "uiId"
                }],
                modal : {
                    move : {
                        id : 1,
                        parentId : 2
                    }
                }
            })
            const document = test(MoveModal, global, global, {
                fetch
            })
            expect(global.ui.uiId).toBeFalsy()
            document.click("move_modal_save_button")
            expect(fetch).toBeCalledWith("/api/file", {
                method : "PATCH",
                headers : {
                    "Content-Type" : "application/json; charset=utf-8"
                },
                body : JSON.stringify({
                    id : 1,
                    parentId : 2
                })
            })
            expect(global.ui.uiId).toBe(true)
        })
    })
})