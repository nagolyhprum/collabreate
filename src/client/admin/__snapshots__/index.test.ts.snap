// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Editor renders 1`] = `
Object {
  "cache": Set {
    "onInit:editor_root",
    "observe:project_name",
    "observe:branch_name",
    "onClick:add_file_button",
    "onClick:add_folder_button",
    "observe:root_folder_file",
    "root_folder_file_folder",
    "onClick:folder_component_expand_collapse_button",
    "observe:folder_component_expand_collapse_text",
    "observe:folder_component_name",
    "onClick:folder_component_rename_button",
    "onClick:folder_component_move_button",
    "onClick:folder_component_remove_button",
    "onClick:folder_component_add_file_button",
    "onClick:folder_component_add_folder_button",
    "observe:folder_component_children",
    "folder_component_children_file",
    "observe:file_component_name",
    "onClick:file_component_rename_buttom",
    "onClick:file_component_move_buttom",
    "onClick:file_component_remove_buttom",
    "folder_component_children_folder",
    "root_folder_file_file",
    "observe:rename_modal",
    "observe:rename_modal_title",
    "observe:rename_modal_name_input",
    "onInput:rename_modal_name_input",
    "onClick:rename_modal_cancel_button",
    "observe:rename_modal_save_button",
    "onClick:rename_modal_save_button",
    "observe:remove_modal",
    "observe:remove_modal_name_input",
    "onInput:remove_modal_name_input",
    "onClick:remove_modal_cancel_button",
    "observe:remove_modal_save_button",
    "onClick:remove_modal_save_button",
    "observe:move_modal",
    "observe:move_modal_title",
    "onSelect:move_modal_parent_select",
    "observe:move_modal_parent_select",
    "move_modal_parent_select_option",
    "observe:standard_select_option",
    "onClick:move_modal_cancel_button",
    "onClick:move_modal_save_button",
  },
  "css": Array [],
  "html": Array [
    "<div style=\\"height:auto;width:auto\\">",
    "<div style=\\"width:100%;height:100%;display:flex;flex-direction:row\\" data-id=\\"editor_root\\">",
    "<div style=\\"width:30%;height:100%;overflow:auto;background:yellow\\">",
    "<div style=\\"width:100%;height:auto;display:flex;flex-direction:column;padding-top:16px;padding-right:16px;padding-bottom:16px;padding-left:16px\\">",
    "<div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\">",
    "<span style=\\"width:auto;height:auto\\" data-id=\\"project_name\\">",
    "",
    "</span>",
    "<span style=\\"width:auto;height:auto\\" data-id=\\"branch_name\\">",
    "",
    "</span>",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"add_file_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Add File",
    "</span>",
    "</button>",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"add_folder_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Add Folder",
    "</span>",
    "</button>",
    "</div>",
    "<div style=\\"width:100%;height:auto;display:flex;flex-direction:column;border-left:1px solid black\\" data-id=\\"root_folder_file\\">",
    "</div>",
    "</div>",
    "</div>",
    "<div style=\\"width:0px;height:100%;display:flex;flex-direction:column;flex-grow:1;background:purple\\">",
    "</div>",
    "<div style=\\"width:100%;height:100%;position:absolute;background:rgba(0, 0, 0, 0.7);top:0px;left:0px;display:none\\" data-id=\\"rename_modal\\">",
    "<div style=\\"width:auto;height:auto;display:flex;flex-direction:column;background:white;padding-top:16px;padding-right:16px;padding-bottom:16px;padding-left:16px;position:absolute;left:50%;top:50%\\">",
    "<span style=\\"width:auto;height:auto\\" data-id=\\"rename_modal_title\\">",
    "Renaming: ",
    "</span>",
    "<input style=\\"width:100%;height:auto\\" data-id=\\"rename_modal_name_input\\"/>",
    "<div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\">",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"rename_modal_cancel_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Cancel",
    "</span>",
    "</button>",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"rename_modal_save_button\\" disabled=\\"disabled\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Save",
    "</span>",
    "</button>",
    "</div>",
    "</div>",
    "</div>",
    "<div style=\\"width:100%;height:100%;position:absolute;background:rgba(0, 0, 0, 0.7);top:0px;left:0px;display:none\\" data-id=\\"remove_modal\\">",
    "<div style=\\"width:auto;height:auto;display:flex;flex-direction:column;background:white;padding-top:16px;padding-right:16px;padding-bottom:16px;padding-left:16px;position:absolute;left:50%;top:50%\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Please confirm the name of the file you would like to remove",
    "</span>",
    "<input style=\\"width:100%;height:auto\\" data-id=\\"remove_modal_name_input\\"/>",
    "<div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\">",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"remove_modal_cancel_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Cancel",
    "</span>",
    "</button>",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"remove_modal_save_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Remove",
    "</span>",
    "</button>",
    "</div>",
    "</div>",
    "</div>",
    "<div style=\\"width:100%;height:100%;position:absolute;background:rgba(0, 0, 0, 0.7);top:0px;left:0px;display:none\\" data-id=\\"move_modal\\">",
    "<div style=\\"width:auto;height:auto;display:flex;flex-direction:column;background:white;padding-top:16px;padding-right:16px;padding-bottom:16px;padding-left:16px;position:absolute;left:50%;top:50%\\">",
    "<span style=\\"width:auto;height:auto\\" data-id=\\"move_modal_title\\">",
    "Where would you like to move ",
    "</span>",
    "<select style=\\"width:100%;height:auto\\" data-id=\\"move_modal_parent_select\\">",
    "<option style=\\"width:auto;height:auto\\" data-id=\\"standard_select_option\\" value=\\"null\\">",
    "<root>",
    "</option>",
    "</select>",
    "<div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\">",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"move_modal_cancel_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Cancel",
    "</span>",
    "</button>",
    "<button style=\\"width:auto;height:auto\\" data-id=\\"move_modal_save_button\\">",
    "<span style=\\"width:auto;height:auto\\">",
    "Move",
    "</span>",
    "</button>",
    "</div>",
    "</div>",
    "</div>",
    "</div>",
    "</div>",
  ],
  "js": Array [
    "var adapters = {};",
    "var events = {};",
    "var listeners = [];",
    "var global = {\\"selectedDirectory\\":\\"components\\",\\"files\\":[],\\"components\\":[],\\"ui\\":{},\\"branch\\":{\\"id\\":\\"\\",\\"previousBranchId\\":\\"\\",\\"projectId\\":\\"\\"},\\"project\\":{\\"id\\":\\"\\",\\"latestBranchId\\":\\"\\"},\\"modal\\":{\\"remove\\":{\\"id\\":-1,\\"name\\":\\"\\",\\"input\\":\\"\\"},\\"rename\\":{\\"id\\":-1,\\"name\\":\\"\\",\\"input\\":\\"\\"},\\"move\\":{\\"id\\":-1,\\"name\\":\\"\\",\\"parentId\\":null}}};",
    "setEvent(\\"editor_root\\", \\"onInit\\", function(local, index, event) {",
    "	socket.on(\\"file.upsert\\", function (args) {
		var data = args.data;
		global.files = _.upsert(global.files, data);
	});
	socket.on(\\"file.remove\\", function (args) {
		var data = args.data;
		global.files = _.map(_.filter(global.files, function (args) {
			var item = args.item;
			return !((item.id === data));
		}), function (args) {
			var item = args.item;
			if((item.parentId === data)) {
				return _.assign(item, {
					parentId: null
				});
			} else {
				return item;
			}
		});
	});
	fetch(\\"/api/project\\", {
		callback: function (args) {
			var body = args.body;
			var json = JSON.parse(body);
			global.project = json.project;
			global.branch = json.branch;
			global.files = json.files;
			global.components = json.components;
		}
	});",
    "});",
    "setEvent(\\"project_name\\", \\"observe\\", function(local, index, event) {",
    "	event.text = global.project.id;",
    "});",
    "setEvent(\\"branch_name\\", \\"observe\\", function(local, index, event) {",
    "	event.text = global.branch.id;",
    "});",
    "setEvent(\\"add_file_button\\", \\"onClick\\", function(local, index, event) {",
    "	fetch(\\"/api/file\\", {
		method: \\"POST\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			branchId: global.branch.id,
			isFolder: false,
			parentId: null
		})
	});",
    "});",
    "setEvent(\\"add_folder_button\\", \\"onClick\\", function(local, index, event) {",
    "	fetch(\\"/api/file\\", {
		method: \\"POST\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			branchId: global.branch.id,
			isFolder: true,
			parentId: null
		})
	});",
    "});",
    "setEvent(\\"root_folder_file\\", \\"observe\\", function(local, index, event) {",
    "	event.data = _.map(_.filter(global.files, function (args) {
		var item = args.item;
		return (item.parentId === null);
	}), function (args) {
		var item = args.item;
		if(item.isFolder) {
			return _.assign(item, {
				adapter: \\"folder\\"
			});
		} else {
			return _.assign(item, {
				adapter: \\"file\\"
			});
		}
	});",
    "});",
    "setEvent(\\"folder_component_expand_collapse_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.ui[local.uiId] = !(global.ui[local.uiId]);",
    "});",
    "setEvent(\\"folder_component_expand_collapse_text\\", \\"observe\\", function(local, index, event) {",
    "	if(global.ui[local.uiId]) {
		event.text = \\"-\\";
	} else {
		event.text = \\"+\\";
	}",
    "});",
    "setEvent(\\"folder_component_name\\", \\"observe\\", function(local, index, event) {",
    "	event.text = local.name;",
    "});",
    "setEvent(\\"folder_component_rename_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.rename = {
		id: local.id,
		name: local.name,
		input: local.name
	};",
    "});",
    "setEvent(\\"folder_component_move_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.move = {
		id: local.id,
		name: local.name,
		parentId: local.parentId
	};",
    "});",
    "setEvent(\\"folder_component_remove_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.remove = {
		id: local.id,
		name: local.name,
		input: \\"\\"
	};",
    "});",
    "setEvent(\\"folder_component_add_file_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.ui[local.uiId] = true;
	fetch(\\"/api/file\\", {
		method: \\"POST\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			branchId: global.branch.id,
			isFolder: false,
			parentId: local.id
		})
	});",
    "});",
    "setEvent(\\"folder_component_add_folder_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.ui[local.uiId] = true;
	fetch(\\"/api/file\\", {
		method: \\"POST\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			branchId: global.branch.id,
			isFolder: true,
			parentId: local.id
		})
	});",
    "});",
    "setEvent(\\"folder_component_children\\", \\"observe\\", function(local, index, event) {",
    "	event.visible = global.ui[local.uiId];",
    "	event.data = _.map(_.filter(global.files, function (args) {
		var item = args.item;
		return (item.parentId === local.id);
	}), function (args) {
		var item = args.item;
		if(item.isFolder) {
			return _.assign(item, {
				adapter: \\"folder\\"
			});
		} else {
			return _.assign(item, {
				adapter: \\"file\\"
			});
		}
	});",
    "});",
    "setEvent(\\"file_component_name\\", \\"observe\\", function(local, index, event) {",
    "	event.text = local.name;",
    "});",
    "setEvent(\\"file_component_rename_buttom\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.rename = {
		id: local.id,
		name: local.name,
		input: local.name
	};",
    "});",
    "setEvent(\\"file_component_move_buttom\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.move = {
		id: local.id,
		name: local.name,
		parentId: local.parentId
	};",
    "});",
    "setEvent(\\"file_component_remove_buttom\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.remove = {
		id: local.id,
		name: local.name,
		input: \\"\\"
	};",
    "});",
    "adapters.folder_component_children_file = $('<div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\"><span style=\\"width:auto;height:auto;padding-top:16px;padding-right:16px;padding-bottom:0px;padding-left:16px\\" data-id=\\"file_component_name\\"></span><button style=\\"width:auto;height:auto\\" data-id=\\"file_component_rename_buttom\\"><span style=\\"width:auto;height:auto\\">Rename</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"file_component_move_buttom\\"><span style=\\"width:auto;height:auto\\">Move</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"file_component_remove_buttom\\"><span style=\\"width:auto;height:auto\\">Remove</span></button></div>')",
    "adapters.folder_component_children_folder = $('<div style=\\"width:100%;height:auto;display:flex;flex-direction:column;padding-top:16px;padding-right:16px;padding-bottom:0px;padding-left:16px\\"><div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\"><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_expand_collapse_button\\"><span style=\\"width:auto;height:auto\\" data-id=\\"folder_component_expand_collapse_text\\">+</span></button><span style=\\"width:auto;height:auto\\" data-id=\\"folder_component_name\\"></span><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_rename_button\\"><span style=\\"width:auto;height:auto\\">Rename</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_move_button\\"><span style=\\"width:auto;height:auto\\">Move</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_remove_button\\"><span style=\\"width:auto;height:auto\\">Remove</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_add_file_button\\"><span style=\\"width:auto;height:auto\\">Add File</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_add_folder_button\\"><span style=\\"width:auto;height:auto\\">Add Folder</span></button></div><div style=\\"width:100%;height:auto;display:none;flex-direction:column;border-left:1px solid black\\" data-id=\\"folder_component_children\\"></div></div>')",
    "adapters.root_folder_file_folder = $('<div style=\\"width:100%;height:auto;display:flex;flex-direction:column;padding-top:16px;padding-right:16px;padding-bottom:0px;padding-left:16px\\"><div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\"><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_expand_collapse_button\\"><span style=\\"width:auto;height:auto\\" data-id=\\"folder_component_expand_collapse_text\\">+</span></button><span style=\\"width:auto;height:auto\\" data-id=\\"folder_component_name\\"></span><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_rename_button\\"><span style=\\"width:auto;height:auto\\">Rename</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_move_button\\"><span style=\\"width:auto;height:auto\\">Move</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_remove_button\\"><span style=\\"width:auto;height:auto\\">Remove</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_add_file_button\\"><span style=\\"width:auto;height:auto\\">Add File</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"folder_component_add_folder_button\\"><span style=\\"width:auto;height:auto\\">Add Folder</span></button></div><div style=\\"width:100%;height:auto;display:none;flex-direction:column;border-left:1px solid black\\" data-id=\\"folder_component_children\\"></div></div>')",
    "adapters.root_folder_file_file = $('<div style=\\"width:100%;height:auto;display:flex;flex-direction:row\\"><span style=\\"width:auto;height:auto;padding-top:16px;padding-right:16px;padding-bottom:0px;padding-left:16px\\" data-id=\\"file_component_name\\"></span><button style=\\"width:auto;height:auto\\" data-id=\\"file_component_rename_buttom\\"><span style=\\"width:auto;height:auto\\">Rename</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"file_component_move_buttom\\"><span style=\\"width:auto;height:auto\\">Move</span></button><button style=\\"width:auto;height:auto\\" data-id=\\"file_component_remove_buttom\\"><span style=\\"width:auto;height:auto\\">Remove</span></button></div>')",
    "setEvent(\\"rename_modal\\", \\"observe\\", function(local, index, event) {",
    "	event.visible = !((global.modal.rename.id === -1));",
    "});",
    "setEvent(\\"rename_modal_title\\", \\"observe\\", function(local, index, event) {",
    "	event.text = (\\"Renaming: \\" + global.modal.rename.name);",
    "});",
    "setEvent(\\"rename_modal_name_input\\", \\"observe\\", function(local, index, event) {",
    "	event.value = global.modal.rename.input;
	event.focus = global.modal.rename.id;",
    "});",
    "setEvent(\\"rename_modal_name_input\\", \\"onInput\\", function(local, index, event) {",
    "	global.modal.rename.input = event;",
    "});",
    "setEvent(\\"rename_modal_cancel_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.rename.id = -1;",
    "});",
    "setEvent(\\"rename_modal_save_button\\", \\"observe\\", function(local, index, event) {",
    "	event.enabled = !((global.modal.rename.input === \\"\\"));",
    "});",
    "setEvent(\\"rename_modal_save_button\\", \\"onClick\\", function(local, index, event) {",
    "	fetch(\\"/api/file\\", {
		method: \\"PATCH\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			id: global.modal.rename.id,
			name: global.modal.rename.input
		})
	});
	global.modal.rename.id = -1;",
    "});",
    "setEvent(\\"remove_modal\\", \\"observe\\", function(local, index, event) {",
    "	event.visible = !((global.modal.remove.id === -1));",
    "});",
    "setEvent(\\"remove_modal_name_input\\", \\"observe\\", function(local, index, event) {",
    "	event.value = global.modal.remove.input;
	event.placeholder = global.modal.remove.name;
	event.focus = global.modal.remove.id;",
    "});",
    "setEvent(\\"remove_modal_name_input\\", \\"onInput\\", function(local, index, event) {",
    "	global.modal.remove.input = event;",
    "});",
    "setEvent(\\"remove_modal_cancel_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.remove.id = -1;",
    "});",
    "setEvent(\\"remove_modal_save_button\\", \\"observe\\", function(local, index, event) {",
    "	event.enabled = (global.modal.remove.input === global.modal.remove.name);",
    "});",
    "setEvent(\\"remove_modal_save_button\\", \\"onClick\\", function(local, index, event) {",
    "	fetch(\\"/api/file\\", {
		method: \\"DELETE\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			id: global.modal.remove.id
		})
	});
	global.modal.remove.id = -1;",
    "});",
    "setEvent(\\"move_modal\\", \\"observe\\", function(local, index, event) {",
    "	event.visible = !((global.modal.move.id === -1));",
    "});",
    "setEvent(\\"move_modal_title\\", \\"observe\\", function(local, index, event) {",
    "	event.text = (\\"Where would you like to move \\" + global.modal.move.name);",
    "});",
    "setEvent(\\"move_modal_parent_select\\", \\"onSelect\\", function(local, index, event) {",
    "	global.modal.move.parentId = JSON.parse(event);",
    "});",
    "setEvent(\\"move_modal_parent_select\\", \\"observe\\", function(local, index, event) {",
    "	event.data = _.concat([{
		adapter: \\"option\\",
		label: \\"<root>\\",
		value: \\"null\\"
	}], _.map(_.filter(global.files, function (args) {
		var item = args.item;
		return (!((item.id === global.modal.move.id)) && item.isFolder);
	}), function (args) {
		var item = args.item;
		return {
			adapter: \\"option\\",
			label: item.name,
			value: item.id
		};
	}));
	event.value = global.modal.move.parentId;",
    "});",
    "setEvent(\\"standard_select_option\\", \\"observe\\", function(local, index, event) {",
    "	event.text = local.label;
	event.value = local.value;",
    "});",
    "adapters.move_modal_parent_select_option = $('<option style=\\"width:auto;height:auto\\" data-id=\\"standard_select_option\\"></option>')",
    "setEvent(\\"move_modal_cancel_button\\", \\"onClick\\", function(local, index, event) {",
    "	global.modal.move.id = -1;",
    "});",
    "setEvent(\\"move_modal_save_button\\", \\"onClick\\", function(local, index, event) {",
    "	fetch(\\"/api/file\\", {
		method: \\"PATCH\\",
		headers: {
			\\"Content-Type\\": \\"application/json; charset=utf-8\\"
		},
		body: JSON.stringify({
			id: global.modal.move.id,
			parentId: global.modal.move.parentId
		})
	});
	global.modal.move.id = -1;",
    "});",
    "bind(document.body, Local(global, 0));",
  ],
}
`;
